File: CarFleet/car_fleet.py
def car_fleet(target: int, position: list[int], speed: list[int]) -> int:
    cars = sorted(zip(position, speed))
    times = []
    for p, s in cars[::-1]:
        time = (target - p) / s
        times.append(time)
        if len(times) >= 2 and times[-1] <= times[-2]:
            times.pop()
    return len(times)

File: CarFleet/CarFleet.hs
import Data.Function (on)
import Data.List (sortBy)

carFleet :: Int -> [Int] -> [Int] -> Int
carFleet target position speed = length $ foldl go [] cars
  where
    cars = sortBy (on (flip compare) fst) $ zip position speed
    go :: [Float] -> (Int, Int) -> [Float]
    go [] (p, s) = [on (/) fromIntegral (target - p) s]
    go (t' : ts) (p, s)
      | t <= t' = t' : ts
      | otherwise = t : t' : ts
      where
        t = on (/) fromIntegral (target - p) s
